### DISCUSSION TITLE
Are dump and parse guaranteed to be inverse to each other?

### DISCUSSION THREAD

**Discussion by Mathias-Fuchs** [nlohmann/json#2788] [Q&A] [answered]
The following example seems to imply that any json can be dumped and re-parsed, and will stay the same. Is this _guaranteed_ to work on _any_ nlohmann::json object?
Serializing stuff is the point of json. So, it would be great if the README.md was clear about this point. So far, the README only speaks about the other direction, parsing and then dumping a string, which is less useful to know.
```cpp
       nlohmann::json j;
        j["k"] = {"v1", "v2"};


        std::string jj = j.dump();
        nlohmann::json jjj = nlohmann::json::parse(jj);

        std::cout << j << std::endl; // {"k":["v1","v2"]}
        std::cout << jjj << std::endl; // couts the same 
```

**Comment by gregmarr** [comment, marked as answer]
I would say no, it wouldn't be guaranteed, since there can be things in the `json` object that can't be represented in a JSON file, such as NAN or infinity, or non-utf8 strings.

**Reply by nlohmann** [reply]
Right - if you take an arbitrary `json` value, then `dump` may throw in case of invalid UTF-8 or lose information (e.g., by serializing NaN to `null`). But even if your `json` value was created by parsing JSON, the result of `dump()` may not result in the same JSON text as indentation or exact number formatting is not preserved.
